python/cudf/cudf/core/join/join.py

```diff
@@ -60,140 +122,245 @@ def __init__(
+                right_keys.extend(
+                    [
+                        _Indexer(name=on, column=True)
+                        for on in _coerce_to_tuple(self.right_on)
+                    ]
+                )
+        else:
+            # Use `on` if provided. Otherwise,
+            # implicitly use identically named columns as the key columns:
+            on_names = (
+                _coerce_to_tuple(self.on)
+                if self.on is not None
+                else set(self.lhs._data) & set(self.rhs._data)
+            )
+            left_keys = [_Indexer(name=on, column=True) for on in on_names]
+            right_keys = [_Indexer(name=on, column=True) for on in on_names]
+
+        if len(left_keys) != len(right_keys):
+            raise ValueError(
+                "Merge operands must have same number of join key columns"
+            )
+
+        self._keys = _JoinKeys(left=left_keys, right=right_keys)
+
+    def _merge_results(self, left_result: Frame, right_result: Frame) -> Frame:
+        # Merge the Frames `left_result` and `right_result` into a single
+        # `Frame`, suffixing column names if necessary.
+
+        # If two key columns have the same name, a single output column appears
+        # in the result. For all other join types, the key column from the rhs
+        # is simply dropped. For outer joins, the two key columns are combined
+        # by filling nulls in the left key column with corresponding values
+        # from the right key column:
+        if self.how == "outer":
+            for lkey, rkey in zip(*self._keys):
+                if lkey.name == rkey.name:
+                    # fill nulls in lhs from values in the rhs
+                    lkey.set(
+                        left_result,
+                        lkey.get(left_result).fillna(rkey.get(right_result)),
+                        validate=False,
+                    )
+
+        # Compute the result column names:
+        # left_names and right_names will be a mappings of input column names
+        # to the corresponding names in the final result.
+        left_names = OrderedDict(zip(left_result._data, left_result._data))
+        right_names = OrderedDict(zip(right_result._data, right_result._data))
+
+        # For any columns from left_result and right_result that have the same
+        # name:
+        # - if they are key columns, keep only the left column
+        # - if they are not key columns, use suffixes to differentiate them
+        #   in the final result
+        common_names = set(left_names) & set(right_names)
+
+        if self.on:
+            key_columns_with_same_name = self.on
+        else:
+            key_columns_with_same_name = []
+            for lkey, rkey in zip(*self._keys):
+                if (lkey.index, rkey.index) == (False, False):
+                    if lkey.name == rkey.name:
+                        key_columns_with_same_name.append(lkey.name)
```

Comment:
Can turn this into a comprehension. Also given the usage below, should this be a set to make the lookups faster?

Reply:
Done